### PR TITLE
fix(firestore-bigquery-export): accept ISO 8601 string partition values

### LINKED ISSUE #2803: 🐛 [firestore-bigquery-export] String partition values silently become NULL after 0.3.0 (regression vs 0.2.x)
### [REQUIRED] Step 2: Describe your configuration

- Extension name: **firestore-bigquery-export**
- Extension version: **0.3.1** (regression appears to have been introduced in 0.3.0 / change-tracker 2.x — see analysis below)
- Configuration values (redacted):
  - `COLLECTION_PATH`: `myCollection`
  - `TIME_PARTITIONING`: `DAY`
  - `TIME_PARTITIONING_FIELD`: `order_week`
  - `TIME_PARTITIONING_FIELD_TYPE`: `DATE`
  - `TIME_PARTITIONING_FIRESTORE_FIELD`: `order_week`

The Firestore field `order_week` holds an ISO 8601 date string (e.g. `"2026-01-01"`), and the BigQuery column type is `DATE`. This worked correctly under 0.2.x because BigQuery streaming insert implicitly casts ISO 8601 date strings to `DATE`.

### [REQUIRED] Step 3: Describe the problem

After upgrading from 0.2.x to 0.3.1, the partition column is written as **`NULL` for every row whose Firestore field is a string** (including valid ISO 8601 date strings such as `"2026-01-01"`). All such rows now end up in the `__NULL__` partition, which breaks partition pruning and existing analytics queries.

The change is **not mentioned in the CHANGELOG**:
- `0.2.11`: only documents `chore: bump firestore-bigquery-change-tracker dependency to v2`
- `0.3.0`: only documents partitioning **config-validation** changes and `NONE` / `omit` sentinel normalization

The CHANGELOG gives no hint that the runtime contract for partition field values has narrowed.

#### Steps to reproduce

1. Create a Firestore collection where each document has a string ISO 8601 date field, e.g. `{ order_week: "2026-01-01" }`
2. Install firestore-bigquery-export `0.3.1` with:
   - `TIME_PARTITIONING=DAY`
   - `TIME_PARTITIONING_FIELD=order_week`
   - `TIME_PARTITIONING_FIELD_TYPE=DATE`
   - `TIME_PARTITIONING_FIRESTORE_FIELD=order_week`
3. Write a document to the collection
4. Inspect the row in `<table>_raw_changelog`

##### Expected result

Top-level `order_week` column equals `2026-01-01`; the row lives in the `2026-01-01` DATE partition. This matches 0.2.x behaviour and is what BigQuery streaming insert supports natively.

##### Actual result

Top-level `order_week` column is `NULL`; the row lives in the `__NULL__` partition. The `data` JSON column still contains the original `order_week` string, confirming the value was not lost upstream — only the partition column extraction drops it.

```sql
SELECT
  document_id,
  order_week,                                     -- top-level partition column: NULL
  JSON_VALUE(data, '$.order_week') AS data_order_week  -- still present
FROM `<project>.<dataset>.<table>_raw_changelog`
WHERE order_week IS NULL
ORDER BY timestamp DESC
LIMIT 5;
```

#### Possible cause (from reading the source)

In **0.2.x**, `Partitioning.getPartitionValue()` appears to accept strings as-is via `isValidPartitionTypeString`, leaving the cast to BigQuery on streaming insert:

https://github.com/firebase/extensions/blob/7f52a39b1945de28f287a796b816e97918ab4402/firestore-bigquery-export/firestore-bigquery-change-tracker/src/bigquery/partitioning.ts#L173-L213

In **0.3.x**, after the partitioning refactor (#2447), `PartitionValueConverter.convert()` seems to only accept `firebase.firestore.Timestamp`, `{ _seconds, _nanoseconds }`, or `Date`, and to return `null` for any other type — which would include ISO 8601 date strings:

https://github.com/firebase/extensions/blob/master/firestore-bigquery-export/firestore-bigquery-change-tracker/src/bigquery/partitioning/converter.ts

`getPartitionValue` then omits the column, which would explain the `NULL` we observe in BigQuery:

https://github.com/firebase/extensions/blob/master/firestore-bigquery-export/firestore-bigquery-change-tracker/src/bigquery/partitioning/index.ts

#### Related

- #1483 — existing feature request to allow ISO 8601 strings (predates this regression; would be effectively closed-by-fix)
- #2447 — the refactor PR that introduced the new converter

### PR DESCRIPTION
**Pull request by cabljac**
## Summary

- Restore 0.2.x behavior where Firestore string fields used as time partition values (e.g. `order_week: "2026-01-01"` with `TIME_PARTITIONING_FIELD_TYPE=DATE`) are accepted. The 0.3.0 refactor narrowed `PartitionValueConverter` to Firestore `Timestamp` / timestamp-like / `Date` only, silently coercing strings to `NULL` and landing rows in the `__NULL__` partition.
- `PartitionValueConverter.convert` now parses strings via `new Date(value)`. Unparseable strings still return `null` and trigger the existing warning log.
- Defensive `try/catch` around the BigQuery formatter switch so any serialization failure degrades to `null` + warn rather than crashing the row write.

Fixes #2803

## Test plan

- [x] `cd firestore-bigquery-export/firestore-bigquery-change-tracker && npm run test:local -- src/__tests__/bigquery/partitioning/converter.test.ts` — 29/29 pass
- [x] Full tracker `npm run test:local` — converter + config + unit suites pass; the 8 pre-existing failures in `e2e.test.ts` are present on `origin/next` baseline (verified) and unrelated to this change
- [x] `npx tsc --noEmit` — clean
- [x] Maintainer to bump `extension.yaml` version to `0.3.2` as part of release workflow (CHANGELOG entry added)

### CODE REVIEW

**Comment by cabljac** [comment]
Thanks for the review. Addressing each point:

**1. Stripping `Z` from DATETIME**: I verified empirically — `BigQuery.datetime("2024-01-15T10:30:00.000Z")` already normalizes to `"2024-01-15 10:30:00.000"` (canonical BQ DATETIME with space separator). Stripping `Z` would actually regress the output to `"2024-01-15T10:30:00.000"` (T-separator passthrough). This same Z-suffixed code path has been shipping since 0.3.0 for Firestore Timestamp / Date inputs, so we'd see existing breakage if the library rejected it.

**2. `default: return null`**: Good catch — applied in d0654faf. TS exhaustiveness covers this at compile time, but `fieldType` comes from external config so a runtime mismatch is worth handling cleanly.

**3. Logging in catch**: The caller `Partitioning.getPartitionValue()` already logs whenever `convert()` returns null ([index.ts L79–L86](https://github.com/firebase/extensions/blob/fix/bq-part/firestore-bigquery-export/firestore-bigquery-change-tracker/src/bigquery/partitioning/index.ts#L79-L86)) — so the existing warn fires for the catch path too. By design, the converter is pure and the caller logs.

**Comment by CorieW** [comment]
The overall direction looks right, but I think this needs one more change before merging.

The current implementation uses `new Date(value)` to validate string partition values. That is risky because JavaScript date parsing is too permissive and can silently normalize invalid or partial dates.

For example:

```ts
new Date("2024-02-30").toISOString(); // Wrong: "2024-02-30" is not a real date, but JS normalizes it to "2024-03-01T00:00:00.000Z".
new Date("2024-01").toISOString();    // Wrong: this is a partial date, but JS treats it as "2024-01-01T00:00:00.000Z".
new Date("1").toISOString();          // Wrong: this is not a date format we should accept, but JS parses it as "2001-01-01T00:00:00.000Z".

**Comment by cabljac** [comment]
Good catch, agreed — `new Date()` alone is too permissive. Verified your three examples plus a couple more silent normalizations on Node 22:

```
"2024-02-30"        -> 2024-03-01  (month overflow)
"2024-01"           -> 2024-01-01  (partial-date fill)
"1"                 -> 2001-01-01  (bare numeric as year)
"2023-02-29"        -> 2023-03-01  (non-leap-year overflow)
"2024-01-15T10:30"  -> engine-dependent (local-time interpretation)
```

Tightened in 4392a4f7:

- **Strict regex** requiring `YYYY-MM-DD` prefix, with optional `T` (or space) followed by `HH:MM[:SS[.ffffff]]` and a **required** timezone designator (`Z` or `±HH:MM`) when the time portion is present. Rejects partial dates, bare numerics, and implicit-local-time datetimes.
- **Calendar validator** built via `setUTCFullYear(y, m-1, d)` then comparing UTC components. Catches Feb 30, non-leap-year Feb 29, month 13, day 32, etc. Uses `setUTCFullYear` rather than `Date.UTC()` to avoid the legacy 2-digit-year quirk for years 1–99.
- Existing pinned tests (date-only, Z-suffixed datetime, timezone-shifted DATE column) still pass; added 10 new rejection tests.

**Comment by cabljac** [comment]
One more gap caught on self-review, fixed in fba4ca3a.

`"0000-01-01"` was passing through all my validation layers:

- regex matches (`\d{4}` accepts `0000`)
- calendar validator passes (`setUTCFullYear(0, 0, 1)` yields year 0, matching the input components)
- `new Date("0000-01-01")` parses fine

But BigQuery DATE / DATETIME / TIMESTAMP all reject year 0 (the supported range is 0001-01-01 to 9999-12-31), so the row would have failed server-side and ended up as the same NULL-partition symptom this PR is meant to fix, just on a more obscure input. Added an explicit `yearN < 1` guard plus tests covering year 0 (rejected), year 0001 (accepted, BQ minimum), and year 9999 (accepted, BQ maximum).

Other edge cases I probed against the regex + validator + parser chain and confirmed are handled:

- hour 25, minute 60, offset > 24h: `new Date()` returns NaN, the `isNaN(parsed.getTime())` check catches
- Feb 30 with a non-UTC offset (e.g. `"2024-02-30T22:00:00-08:00"`): calendar validator catches the Y/M/D mismatch independent of timezone
- single-digit month like `"2024-1-15"`: regex `\d{2}` requires two digits, rejected
- comma fractional separator (`"...,123Z"`): regex requires `.`, rejected

41/41 converter tests passing.

File: firestore-bigquery-export/firestore-bigquery-change-tracker/src/bigquery/partitioning/converter.ts [hunk 1]
```diff
@@ -27,17 +27,61 @@ export class PartitionValueConverter {
...
+    } else if (typeof value === "string") {
+      // Strict ISO 8601 / RFC 3339: YYYY-MM-DD, optionally followed by T or
+      // space-separated HH:MM[:SS[.ffffff]] and a required timezone designator
+      // when the time component is present. JS Date parsing alone is too
+      // permissive — it silently normalizes invalid inputs (e.g. "2024-02-30"
+      // → "2024-03-01"), accepts partial dates ("2024-01"), and reads bare
+      // numerics as years ("1" → "2001-01-01"). Reject all of those.
+      const m = value.match(
+        /^(\d{4})-(\d{2})-(\d{2})(?:[T ](\d{2}):(\d{2})(?::(\d{2})(?:\.\d+)?)?(Z|[+-]\d{2}:?\d{2}))?$/
+      );
```
[the remark this answers is not in the document]
Comment (cabljac) [reply firestore-bigquery-export/firestore-bigquery-change-tracker/src/bigquery/partitioning/converter.ts:39]: Skipping this. V8 caches the compiled regex literal internally, so the runtime cost of the inline form is effectively the cost of the String.prototype.match call itself, not a fresh RegExp compile per invocation. The regex is also only used in one place, so hoisting does not deduplicate anything, and the inline form keeps the pattern next to its explanatory comment block.
